pstoremgr/pstoremgr.go

```diff
@@ -221,13 +236,95 @@ func (pm *Manager) SavePeerstore(addrs []api.Multiaddr) {
 	}
 	defer f.Close()
 
-	for _, a := range addrs {
-		f.Write([]byte(fmt.Sprintf("%s\n", a.Value().String())))
+	for _, pinfo := range pinfos {
+		addrs, err := peerstore.InfoToP2pAddrs(&pinfo)
+		if err != nil {
+			logger.Warning(err)
+			continue
+		}
+		for _, a := range addrs {
+			f.Write([]byte(fmt.Sprintf("%s\n", a.String())))
+		}
 	}
 }
 
-// SavePeerstoreForPeers calls PeersAddresses and then saves the peerstore
+// SavePeerstoreForPeers calls PeerInfos and then saves the peerstore
 // file using the result.
 func (pm *Manager) SavePeerstoreForPeers(peers []peer.ID) {
-	pm.SavePeerstore(pm.PeersAddresses(peers))
+	pm.SavePeerstore(pm.PeerInfos(peers))
+}
+
+// Bootstrap attempts to get as much as count connected peers by selecting
+// randomly from those in the libp2p host peerstore. It returns the number
+// if peers it sucessfully connected to.
+func (pm *Manager) Bootstrap(count int) int {
+	knownPeers := pm.host.Peerstore().PeersWithAddrs()
+	toSort := &peerSort{
+		pinfos: peerstore.PeerInfos(pm.host.Peerstore(), knownPeers),
+		pstore: pm.host.Peerstore(),
+	}
+
+	// Sort from highest to lowest priority
+	sort.Sort(toSort)
+
+	pinfos := toSort.pinfos
+	lenKnown := len(pinfos)
+	totalConns := 0
+
+	// keep conecting while we have peers in the store
+	// and we have not reached count.
+	for i := 0; i < lenKnown && totalConns < count; i++ {
+		pinfo := pinfos[i]
+		ctx, cancel := context.WithTimeout(pm.ctx, ConnectTimeout)
+		defer cancel()
+
+		logger.Infof("connecting to %s", pinfo.ID)
+		err := pm.host.Connect(ctx, pinfo)
+		if err != nil {
+			logger.Warning(err)
+			pm.SetPriority(pinfo.ID, 9999)
+			continue
+		}
+		totalConns++
+	}
+	return totalConns
+}
+
+// SetPriority attaches a priority to a peer. 0 means more priority than
+// 1. 1 means more priority than 2 etc.
+func (pm *Manager) SetPriority(pid peer.ID, prio int) error {
+	return pm.host.Peerstore().Put(pid, PriorityTag, prio)
+}
+
+type peerSort struct {
+	pinfos []peerstore.PeerInfo
+	pstore peerstore.Peerstore
+}
+
+func (ps *peerSort) Len() int {
+	return len(ps.pinfos)
+}
+
+func (ps *peerSort) Less(i, j int) bool {
+	pinfo1 := ps.pinfos[i]
+	pinfo2 := ps.pinfos[j]
+
+	var prio1, prio2 int
+
+	prio1iface, err := ps.pstore.Get(pinfo1.ID, PriorityTag)
+	if err == nil {
+		prio1 = prio1iface.(int)
+	}
+	prio2iface, err := ps.pstore.Get(pinfo2.ID, PriorityTag)
+	if err == nil {
+		prio2 = prio2iface.(int)
+	}
+	return prio1 < prio2
```

Comment:
A peer without priority has the highest priority (0). This is on purpose.

Reply:
This should probably be a comment in the code